src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/RenameTypeService.cs

```diff
@@ -0,0 +1,239 @@
+            codeModel = null;
+            Document newDocument = GetDocument(project, newFilePath);
+            if (newDocument is null)
+                return false;
+
+            IVsHierarchy hierarchy = _workspace.GetHierarchy(project.Id);
+            if (hierarchy is null)
+                return false;
+
+            if (ErrorHandler.Failed(hierarchy.GetProperty(VSConstants.VSITEMID_ROOT, (int)__VSHPROPID.VSHPROPID_ExtObject, out object value)))
+                return false;
+
+            var dteProject = (EnvDTE.Project)value;
+
+            EnvDTE.ProjectItem projectItemForDocument = dteProject.ProjectItems
+                .OfType<EnvDTE.ProjectItem>()
+                .FirstOrDefault(p => StringComparer.InvariantCultureIgnoreCase.Compare(p.Name, newDocument.Name) == 0);
+
+            codeModel = projectItemForDocument.FileCodeModel;
+            return true;
+        }
+
+        private bool TryGetProjectAtPath(string fullPath, out Project project)
+        {
+            // NOTE: It is safe to grab the first project we encounter which
+            // contains the given file. Roslyn will handle the case where the
+            // file is included in multiple projects (linked-files or multi-TFM)
+            foreach (Project proj in _workspace.CurrentSolution.Projects)
+            {
+                if (StringComparers.Paths.Equals(proj.FilePath, fullPath))
+                {
+                    project = proj;
+                    return true;
+                }
+
+            }
+
+            project = null;
+            return false;
+        }
+
+        private async Task<bool> CanHandleRenameAsync(string oldName, string newName, Project project)
+        {
+            (bool success, bool isCaseSensitive) = await TryDetermineIfCompilationIsCaseSensitiveAsync(project);
+            if (!success)
+                return false;
+
+            return IsValidIdentifier(oldName) &&
+                   IsValidIdentifier(newName) &&
+                       (!string.Equals(
+                            oldName,
+                            newName,
+                            isCaseSensitive
+                                ? StringComparison.Ordinal
+                                : StringComparison.OrdinalIgnoreCase));
+        }
+
+        private bool IsValidIdentifier(string identifierName)
+        {
+            return SyntaxFactsService?.IsValidIdentifier(identifierName) ?? false;
+        }
+
+        private static async Task<(bool success, bool isCaseSensitive)> TryDetermineIfCompilationIsCaseSensitiveAsync(Project project)
+        {
+            Compilation compilation = await project.GetCompilationAsync();
+            if (compilation is null)
+            {
+                // this project does not support compilations
+                return (false, false);
+            }
+
+            return (true, compilation.IsCaseSensitive);
+        }
+
+        private static bool  TryGetCodeElementToRename(string oldName,
```

Comment:
You already had the ISymbol from earlier; why do the rename through CodeModel when you can just call the symbol renaming API directly?

Reply:
I don't want to be responsible for calling `TryOnBeforeGlobalSymbolRenamed`/`TryOnAfterGlobalSymbolRenamed`